src/cdn.rs

```diff
@@ -0,0 +1,215 @@
+use crate::Config;
+use anyhow::{Error, Result};
+use aws_sdk_cloudfront::{
+    model::{InvalidationBatch, Paths},
+    Client, RetryConfig,
+};
+use std::sync::{Arc, Mutex};
+use strum::EnumString;
+use tokio::runtime::Runtime;
+use uuid::Uuid;
+
+#[derive(Debug, EnumString)]
+pub(crate) enum CdnKind {
+    #[strum(ascii_case_insensitive)]
+    Dummy,
+
+    #[strum(ascii_case_insensitive)]
+    CloudFront,
+}
+
+pub enum CdnBackend {
+    Dummy(Arc<Mutex<Vec<(String, String)>>>),
+    CloudFront { runtime: Arc<Runtime> },
+}
+
+impl CdnBackend {
+    pub fn new(config: &Arc<Config>, runtime: &Arc<Runtime>) -> CdnBackend {
+        match config.cdn_backend {
+            CdnKind::CloudFront => Self::CloudFront {
+                runtime: runtime.clone(),
+            },
+            CdnKind::Dummy => Self::Dummy(Arc::new(Mutex::new(Vec::new()))),
+        }
+    }
+    /// create a Front invalidation request for a list of path patterns.
+    /// patterns can be
+    /// * `/filename.ext` (a specific path)
+    /// * `/directory-path/file-name.*` (delete these files, all extensions)
+    /// * `/directory-path/*` (invalidate all of the files in a directory, without subdirectories)
+    /// * `/directory-path*` (recursive directory delete, including subdirectories)
+    /// see https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/Invalidation.html#invalidation-specifying-objects
+    ///
+    /// Returns the caller reference that can be used to query the status of this
+    /// invalidation request.
+    pub(crate) fn create_invalidation(
+        &self,
+        distribution_id: &str,
+        path_patterns: &[&str],
+    ) -> Result<Uuid, Error> {
+        let caller_reference = Uuid::new_v4();
+
+        match *self {
+            CdnBackend::CloudFront { ref runtime } => {
+                let shared_config = runtime.block_on(aws_config::load_from_env());
```

Comment:
Should the configuration be loaded during initialization, to avoid running with an invalid config?

Reply:
moved the config & client initialization into the startup, now it matches the S3 backend